File: platform/android/MapToolsApp/src/main/java/com/microsoft/maps/v9/toolsapp/FilterChain.kt
package com.microsoft.maps.v9.toolsapp

import org.maplibre.android.maps.MapLibreMap
import org.maplibre.android.style.expressions.Expression
import org.maplibre.android.style.layers.CircleLayer
import org.maplibre.android.style.layers.FillExtrusionLayer
import org.maplibre.android.style.layers.FillLayer
import org.maplibre.android.style.layers.HeatmapLayer
import org.maplibre.android.style.layers.HillshadeLayer
import org.maplibre.android.style.layers.Layer
import org.maplibre.android.style.layers.LineLayer
import org.maplibre.android.style.layers.PropertyValue
import org.maplibre.android.style.layers.RasterLayer
import org.maplibre.android.style.layers.SymbolLayer

class FilterCondition(val pattern: String, val expression: String? = null) {}
data class FilterExpression(val expression: String, val negative: Boolean)
data class FilterLayer(val id: String, val type: String, val sourceLayer: String)
data class LayerFilterState(val id: String, val visible: Boolean, val expression: String?)

class Filter(val chain: FilterChain, condition: FilterCondition, val negative: Boolean?) {

    val pattern: String = condition.pattern
    val conditions = mutableListOf<FilterExpression>()

    init {
        conditions.add(FilterExpression(condition.expression.orEmpty(), negative ?: false))
    }

    fun combine(condition: FilterCondition, negative: Boolean?) {
        if (pattern != condition.pattern) {
            throw RuntimeException("condition pattern mismatched")
        }
        conditions.add(FilterExpression(condition.expression.orEmpty(), negative ?: false))
    }

    private fun isLayerMatched(lf: FilterLayer): Boolean {
        val test: (String, String) -> Boolean =
            { v, p -> p.isEmpty() || v == p || Regex(p).matches(v) }
        val (layerIdPattern, sourceLayerPattern) = "$pattern;".split(";").take(2)
        return test(lf.id, layerIdPattern) && test(lf.sourceLayer, sourceLayerPattern)
    }

    fun doFilter(fl: FilterLayer): Boolean {
        val map = chain.map ?: throw RuntimeException("map instance not been set")
        if (!isLayerMatched(fl)) {
            return false
        }
        val layer = map.style?.getLayer(fl.id)
            ?: throw RuntimeException("The specified layer#${fl.id} could not found.")
        val dlfs = FilterChain.defaultLayerFilters
        var dlf = dlfs[fl.id]
        if (dlf == null) {
            val visible = layer.visibility.getValue() != "none"
            val expression = layer.getFilterExpression()
            val exp = expression?.toString() ?: ""
            dlf = LayerFilterState(fl.id, visible, exp)
            dlfs[layer.id] = dlf
        }
        if (layer.isVectorLayer()) {
            // Some layers supports filter, so all conditions need to be applied.
            val positives = mutableListOf<Expression>()
            val negatives = mutableListOf<Expression>()
            for (condition in conditions) {
                val exp =
                    if (condition.expression.isNotEmpty()) Expression.Converter.convert(condition.expression) else null
                if (condition.negative) {
                    negatives.add(if (exp != null) Expression.not(exp) else Expression.literal(false))
                } else {
                    positives.add(exp ?: Expression.literal(true))
                }
            }
            var expression: Expression? = null
            if (positives.isNotEmpty()) {
                expression = if (positives.size == 1) positives[0] else Expression.any(*positives.toTypedArray())
            }
            if (negatives.isNotEmpty()) {
                if (expression != null) {
                    negatives.add(expression)
                }
                expression = if (negatives.size == 1) negatives[0] else Expression.all(*negatives.toTypedArray())
            }
            if (!dlf.expression.isNullOrEmpty()) {
                val defaultExpression = Expression.Converter.convert(dlf.expression!!)
                expression = if (expression != null) Expression.all(defaultExpression, expression) else defaultExpression
            }
            layer.setFilterExpression(expression ?: Expression.literal(true))
        } else {
            // Some layers don't support filter and can only be set to visible or hidden, so only the last condition needs to be applied.
            val lastCondition = conditions.last()
            layer.setProperties(
                PropertyValue("visibility", if (lastCondition.negative) "none" else "visible")
            )
        }
        return true
    }
}

class FilterChain(condition: FilterCondition? = null, negative: Boolean? = null) {
    companion object {
        val defaultLayerFilters = mutableMapOf<String, LayerFilterState>()
        var defaultVisibleLayers = listOf<FilterLayer>()

        fun resetAfterStyleChanged() {
            defaultVisibleLayers = listOf<FilterLayer>()
            defaultLayerFilters.clear()
        }
    }

    private val filters = ArrayList<Filter>()
    var map: MapLibreMap? = null

    init {
        if (condition != null) {
            filters.add(Filter(this, condition, negative))
        }
    }

    fun addCondition(condition: FilterCondition, negative: Boolean?): FilterChain {
        filters.find { it.pattern == condition.pattern }?.combine(condition, negative)
            ?: filters.add(Filter(this, condition, negative))
        return this
    }

    fun addConditions(conditions: List<FilterCondition>, negative: Boolean?): FilterChain {
        for (condition in conditions) {
            addCondition(condition, negative)
        }
        return this
    }

    @OptIn(ExperimentalStdlibApi::class)
    private fun getDefaultVisibleLayers(): List<FilterLayer> {
        val map = this.map ?: throw RuntimeException("map instance not been set")

        if (defaultVisibleLayers.isEmpty()) {
            val backgroundLayerId = "microsoft.bing.maps.base.land"

            // Saves the IDs of layers that are displayed by default, which is used to restore the default state of
            // layers that should not be affected when switching filters.
            defaultVisibleLayers = map.style?.layers
                ?.filter { it.id != backgroundLayerId && it.visibility.getValue() != "none" }
                ?.map { FilterLayer(it.id, it.javaClass.simpleName, it.getSourceLayer()) }
                ?: listOf()
        }
        return defaultVisibleLayers
    }

    fun doFilter(map: MapLibreMap) {
        this.map = map
        val layers = getDefaultVisibleLayers()

        // Apply the filters to the layers.
        val layerIdsUpdated = mutableListOf<String>()
        for (filter in filters) {
            layers.filter { lf -> filter.doFilter(lf) }.forEach { layerIdsUpdated.add(it.id) }
        }
        // Reset the layers those are not effected.
        val dlfs = defaultLayerFilters
        for (fl in layers.filter { !layerIdsUpdated.contains(it.id) }) {
            val dlf = dlfs[fl.id] ?: continue
            val layer = map.style?.layers?.find { it.id == fl.id } ?: throw RuntimeException("The specified layer#${fl.id} could not found.")
            if (layer.isVectorLayer()) {
                val exp = if (dlf.expression.isNullOrEmpty()) Expression.literal(true)
                else Expression.Converter.convert(dlf.expression)
                layer.setFilterExpression(exp)
            } else {
                layer.setProperties(
                    PropertyValue("visibility", if (!dlf.visible) "none" else "visible")
                )
            }
        }
    }
}

val filterConditions = mapOf(
    "default" to listOf(FilterCondition("microsoft\\.bing.maps\\..*")),
    "land" to listOf(FilterCondition("microsoft.bing.maps.baseFeature.vector_land")),
    "base" to listOf(FilterCondition("microsoft\\.bing.maps\\.baseFeature\\.[\\w\\d\\-_]+;")),
    // "reserve" to listOf(FilterCondition("microsoft\\.bing\\.maps\\.(baseFeature|labels)\\.[\\w\\d\\-_]+;reserve|golf_course")),
    "reserve" to listOf(FilterCondition("microsoft\\.bing\\.maps\\.baseFeature\\.[\\w\\d\\-_]+;reserve|golf_course")),
    "buildings" to listOf(FilterCondition("microsoft.bing.maps.buildings.buildings")),
    "hillShading" to listOf(FilterCondition("microsoft\\.bing\\.maps\\.hillShading\\.hillShading;")),

    "water" to listOf(FilterCondition("microsoft.bing.maps.baseFeature.generic_water_feature_fill")),
    "sea" to listOf(FilterCondition("microsoft.bing.maps.baseFeature.generic_water_feature_fill", "[\"!\",[\"has\",\"st-et\"]]")),
    "lake" to listOf(FilterCondition("microsoft.bing.maps.baseFeature.generic_water_feature_fill", "[\"==\",[\"get\",\"st-et\"],\"lake\"]")),
    "river" to listOf(FilterCondition("microsoft.bing.maps.baseFeature.generic_water_feature_fill", "[\"==\",[\"get\",\"st-et\"],\"river\"]")),
    "waterName" to listOf(FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_water_feature_(labelonly|polygonlabel);")),
    "seaName" to listOf(FilterCondition("microsoft.bing.maps.labels.generic_water_feature_labelonly;")),
    "lakeName" to listOf(FilterCondition("microsoft.bing.maps.labels.generic_water_feature_polygonlabel", "[\"==\",[\"get\",\"st-et\"],\"lake\"]")),
    "riverName" to listOf(FilterCondition("microsoft.bing.maps.labels.generic_water_feature_polygonlabel", "[\"==\",[\"get\",\"st-et\"],\"river\"]")),

    "continentName" to listOf(FilterCondition("microsoft.bing.maps.labels.entity_override_continents_for_cn_region_symbol_label")),

    "countryRegion" to listOf(FilterCondition("microsoft\\.bing\\.maps\\.roads\\.[\\w\\d\\-_]+;country_region")),
    "countryRegionName" to listOf(FilterCondition("microsoft\\.bing\\.maps\\.labels\\.[\\w\\d\\-_]+;country_region")),

    "island" to listOf(FilterCondition("microsoft\\.bing\\.maps\\.baseFeature\\.[\\w\\d\\-_]+;island")),
    "islandName" to listOf(FilterCondition("microsoft\\.bing\\.maps\\.labels\\.[\\w\\d\\-_]+;island")),

    "cityName" to listOf(
        // 首都
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_(sov_capital|beijing)_(labelonly|iconlabel);"),
        // 一般地名
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)"),
        // 乡镇
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_neighborhood_labelonly;"),
        // 台北
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_taipei_(labelonly|iconlabel)"),
        // 桃园
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_taoyuan_(labelonly|iconlabel)"),
        // 新北
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_new_taipei_(labelonly|iconlabel)"),
        // 其它增补城市（已确定：丹东、东港市、二连浩特）
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_label_orientation_[\\w\\d\\-_]+_(labelonly|iconlabel)"),
    ),
    "capitalCityName" to listOf(
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_(sov_capital|beijing)_(labelonly|iconlabel);"),
        // 其它增补城市（已确定：加德满都、伊斯兰堡）
        FilterCondition(
            "microsoft\\.bing\\.maps\\.labels\\.generic_label_orientation_[\\w\\d\\-_]+_(labelonly|iconlabel)",
            "[\"==\",[\"get\",\"cn-ppl\"],\"sovcap\"]"
        ),
    ),
    "admin1CityName" to listOf(
        FilterCondition(
            "microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)",
            "[\"==\",[\"get\",\"cn-ppl\"],\"admin1cap\"]"
        ),
        // 台北市
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_taipei_(labelonly|iconlabel)"),
    ),
    "admin2CityName" to listOf(
        FilterCondition(
            "microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)",
            "[\"==\",[\"get\",\"cn-ppl\"],\"admin2cap\"]"
        ),
        // 桃园
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_taoyuan_(labelonly|iconlabel)"),
        // 新北
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_new_taipei_(labelonly|iconlabel)"),
        // 其它增补城市（已确定：丹东）
        FilterCondition(
            "microsoft\\.bing\\.maps\\.labels\\.generic_label_orientation_[\\w\\d\\-_]+_(labelonly|iconlabel)",
            "[\"==\",[\"get\",\"cn-ppl\"],\"admin2cap\"]"
        ),

        ),
    // 一般城市（境外的直辖市、大城市等. NOTE: 国内无这个分类!）
    "majorCityName" to listOf(
        FilterCondition(
            "microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)",
            "[\"==\",[\"get\",\"cn-ppl\"],\"city\"]"
        ),
        // 其它增补城市（已确定：东港市、二连浩特）
        FilterCondition(
            "microsoft\\.bing\\.maps\\.labels\\.generic_label_orientation_[\\w\\d\\-_]+_(labelonly|iconlabel)",
            "[\"==\",[\"get\",\"cn-ppl\"],\"city\"]"
        ),
    ),
    "minorCityName" to listOf(
        FilterCondition(
            "microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)",
            "[\"==\",[\"get\",\"cn-ppl\"],\"minorcity\"]"
        ),
        // 其它增补城市（已确定：东港市、二连浩特）
        FilterCondition(
            "microsoft\\.bing\\.maps\\.labels\\.generic_label_orientation_[\\w\\d\\-_]+_(labelonly|iconlabel)",
            "[\"==\",[\"get\",\"cn-ppl\"],\"minorcity\"]"
        ),
    ),
    "townName" to listOf(
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.generic_neighborhood_labelonly;"),
    ),
    "villageName" to listOf(
        FilterCondition(
            "microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)",
            "[\"==\",[\"get\",\"cn-ppl\"],\"village\"]"
        ),
    ),
    "road" to listOf(
        FilterCondition("microsoft\\.bing\\.maps\\.roads\\.[\\w\\d\\-_]+;road[\\w\\d\\-_]*|railway[\\w\\d\\-_]*|tramway"),
    ),
    "railway" to listOf(
        FilterCondition("microsoft\\.bing\\.maps\\.roads\\.[\\w\\d\\-_]+;railway[\\w\\d\\-_]*"),
    ),
    "roadName" to listOf(
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.[\\w\\d\\-_]+;road[\\w\\d\\-_]*|railway[\\w\\d\\-_]*|tramway"),
    ),
    "railwayName" to listOf(
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\.[\\w\\d\\-_]+;railway[\\w\\d\\-_]*"),
    ),
    "roadShield" to listOf(
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\..+_road_shield.*;road"),
    ),
    "landmark" to listOf(
        FilterCondition("microsoft\\.bing\\.maps\\.labels\\..+_landmark(_\\w+)?"),
    ),
)

File: platform/ios/MapToolsApp/MapToolsApp/FilterChain.m
//
//  FilterChain.m
//  MapToolsApp
//
//  Created by Hongbo Yu on 2024/11/12.
//

#import "FilterChain.h"
#import <MapLibre/MLNMapView.h>
#import <MapLibre/MLNBackgroundStyleLayer.h>
#import <MapLibre/MLNSymbolStyleLayer.h>

@implementation NSObject (Internal)

-(NSString*)toJsonStringPrettyPrinted:(BOOL)prettyPrinted {
    NSData* data = [NSJSONSerialization dataWithJSONObject:self options:prettyPrinted ? NSJSONWritingPrettyPrinted : 0 error:nil];
    return data ? [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding] : nil;
}

@end

@implementation NSArray (Internal)
- (id)findObjectUsingBlock:(BOOL (NS_NOESCAPE ^)(id))block {
    for (id it in self) if (block(it)) return it;
    return nil;
}
@end

@interface VisibleLayer: NSObject

@property(nonatomic, readonly) NSString* layerId;
@property(nonatomic, readonly) NSString* type;
@property(nonatomic, readonly) NSString* sourceLayer;
@property(nonatomic, readonly) id filterExpression;

+(instancetype)layerWithLayerId:(NSString*)layerId type:(NSString*)type sourceLayer:(NSString*)sourceLayer filterExpression:(id)filterExpression;
-(instancetype)initWithLayerId:(NSString*)layerId type:(NSString*)type sourceLayer:(NSString*)sourceLayer filterExpression:(id)filterExpression;

@end

@implementation VisibleLayer

+(instancetype)layerWithLayerId:(NSString*)layerId type:(NSString*)type sourceLayer:(NSString*)sourceLayer filterExpression:(id) filterExpression {
    return [[VisibleLayer alloc] initWithLayerId:layerId type:type sourceLayer:sourceLayer filterExpression:filterExpression];
}
-(instancetype)initWithLayerId:(NSString*)layerId type:(NSString*)type sourceLayer:(NSString*)sourceLayer filterExpression:(id)filterExpression {
    self = [super init];
    if (self) {
        _layerId = layerId;
        _type = type;
        _sourceLayer = sourceLayer;
        _filterExpression = filterExpression;
    }
    return self;
}


@end

@implementation FilterCondition

+(instancetype)conditionWithPattern:(NSString*)pattern {
    return [[FilterCondition alloc] initWithPattern:pattern expression:nil];
}

+(instancetype)conditionWithPattern:(NSString*)pattern expression:(NSString*)expression {
    return [[FilterCondition alloc] initWithPattern:pattern expression:expression];
}

-(instancetype)initWithPattern:(NSString*)pattern expression:(nullable NSString*)expression {
    self = [super init];
    if (self) {
        _pattern = pattern;
        _expression = expression;
    }
    return self;
}

@end

@implementation FilterExpression

+(instancetype)expression:(NSString*)expression negative:(BOOL)negative {
    return  [[FilterExpression alloc] initWithExpression:expression negative:negative];
}

-(instancetype)initWithExpression:(NSString*)expression negative:(BOOL)negative {
    self = [super init];
    if (self) {
        _expression = expression;
        _negative = negative;
    }
    return self;
}

@end

@interface FilterChain (Private)
-(NSArray<VisibleLayer*>*)defaultVisibleMapLayers;
@end

@implementation Filter

+(instancetype)filterWithChain:(FilterChain*)chain condition:(FilterCondition*)condition {
    return [[Filter alloc] initWithChain:chain condition:condition negative:NO];
}

+(instancetype)filterWithChain:(FilterChain*)chain condition:(FilterCondition*)condition negative:(BOOL)negative {
    return [[Filter alloc] initWithChain:chain condition:condition negative:negative];
}

-(instancetype)initWithChain:(FilterChain*)chain condition:(FilterCondition*)condition negative:(BOOL)negative {
    self = [super init];
    if (self) {
        _chain = chain;
        _pattern = condition.pattern;
        _conditions = [NSMutableArray arrayWithObject:[FilterExpression expression:condition.expression negative:negative]];
    }
    return self;
}

-(void)combineCondition:(FilterCondition*)condition {
    [self combineCondition:condition negative:NO];
}

-(void)combineCondition:(FilterCondition*)condition negative:(BOOL)negative {
    if (![condition.pattern isEqualToString:_pattern]) {
        @throw [NSException exceptionWithName:@"Combine condition failed" reason:@"Pattern mismatch" userInfo:nil];
    }
    [_conditions addObject:[FilterExpression expression:condition.expression negative:negative]];
}

-(BOOL)isLayerMatchedWithLayerId:(NSString*)layerId sourceLayer:(NSString*)sourceLayer {
    BOOL (^test)(NSString*, NSString*) = ^BOOL(NSString* v, NSString* p) {
        if (![p length] || [p isEqualToString:v]) return YES;
        NSError* error = nil;
        NSRegularExpression* regex = [NSRegularExpression regularExpressionWithPattern:p options:NSRegularExpressionAnchorsMatchLines error:&error];
        NSTextCheckingResult* match = [regex firstMatchInString:v options:0 range:NSMakeRange(0, [v length])];
        return match && match.range.length == v.length;
    };
    NSArray<NSString*>* patterns = [[self->_pattern stringByAppendingString:@";"] componentsSeparatedByString:@";"];
    NSString* layerIdFilter = patterns.count > 0 ? patterns[0] : @"";
    NSString* sourceLayerFilter = patterns.count > 1 ? patterns[1] : @"";
    return test(layerId, layerIdFilter) && test(sourceLayer, sourceLayerFilter);
}

-(BOOL)doFilterWithLayerId:(NSString*)layerId type:(NSString*)type sourceLayer:(NSString*)sourceLayer {
    MLNMapView* mapView = [_chain mapView];
    NSAssert(mapView, @"Map instance not been set.");
    if (![self isLayerMatchedWithLayerId:layerId sourceLayer:sourceLayer]) {
        return NO;
    }
    
    VisibleLayer* visibileLayer = [[_chain defaultVisibleMapLayers] findObjectUsingBlock:^BOOL(VisibleLayer* obj) {
        return [obj.layerId isEqualToString:layerId];
    }];
    MLNStyleLayer* layer = visibileLayer ? [mapView.style layerWithIdentifier:visibileLayer.layerId] : nil;
    NSAssert(layer, @"the layer could not be found.");
    if (!layer) {
        return NO;
    }
    if ([layer isKindOfClass:MLNVectorStyleLayer.class]) {
        NSMutableArray<id>* positiveList = [NSMutableArray array];
        NSMutableArray<id>* negativeList = [NSMutableArray array];
        
        for (FilterExpression* condition in _conditions) {
            id exp = [condition.expression length] ? [NSJSONSerialization JSONObjectWithData:[condition.expression dataUsingEncoding:NSUTF8StringEncoding] options:0 error:nil] : nil;
            if (condition.negative) {
                [negativeList addObject:exp ? @[@"!", exp] : [NSNumber numberWithBool:NO]];
            } else {
                [positiveList addObject:(exp ? exp : [NSNumber numberWithBool:YES])];
            }
        }
        id expression = nil;
        if ([positiveList count]) {
            expression = [positiveList count] == 1 ? positiveList[0] : [@[@"any"] arrayByAddingObjectsFromArray:positiveList];
        }
        if ([negativeList count]) {
            if (expression) {
                [negativeList addObject:expression];
            }
            expression = [negativeList count] == 1 ? negativeList[0] : [@[@"all"] arrayByAddingObjectsFromArray:negativeList];
        }
        id defaultExpression = visibileLayer.filterExpression;
        if (defaultExpression != nil) {
            expression = expression ? @[@"all", defaultExpression, expression] : defaultExpression;
        }
        NSLog(@"layerId=%@", layerId);
        if ([expression isKindOfClass:NSNumber.class]) {
            expression = @[@"all", expression];
        }
        NSLog(@"expression=%@", [expression toJsonStringPrettyPrinted:YES]);

        [((MLNVectorStyleLayer*)layer) setFilterExpression:expression];
    } else {
        FilterExpression* lastCondition = [_conditions lastObject];
        layer.visible = !lastCondition.negative;
        NSLog(@"layerId=%@", layerId);
        NSLog(@"visible=%d", !lastCondition.negative);
    }
    
    return YES;
}

@end


@implementation FilterChain

+(instancetype)chain {
    return [[FilterChain alloc] init];
}

-(instancetype) init {
    self = [super init];
    if (self) {
        _filters = [NSMutableArray array];
    }
    return self;
}

+(instancetype)chainWithCondition:(FilterCondition*)condition {
    return [[FilterChain alloc] initWithCondition:condition];
}

+(instancetype)chainWithCondition:(FilterCondition*)condition negative:(BOOL)negative {
    return [[FilterChain alloc] initWithCondition:condition negative:negative];
}

-(instancetype)initWithCondition:(FilterCondition*)condition {
    return [self initWithCondition:condition negative:NO];
}

-(instancetype)initWithCondition:(FilterCondition*)condition negative:(BOOL)negative {
    self = [super init];
    if (self) {
        _filters = [NSMutableArray arrayWithObject:[Filter filterWithChain:self condition:condition negative:negative]];
    }
    return self;
}

-(void)addCondition:(FilterCondition*)condition {
    [self addCondition:condition negative:NO];
}

-(void)addCondition:(FilterCondition*)condition negative:(BOOL)negative {
    Filter* filter = [_filters findObjectUsingBlock:^BOOL(Filter* obj) {
        return [obj.pattern isEqualToString:condition.pattern];
    }];
    if (filter) {
        [filter combineCondition:condition negative:negative];
    } else {
        [_filters addObject:[Filter filterWithChain:self condition:condition negative:negative]];
    }
}

-(void)doFilterWithMapView:(MLNMapView*)mapView {
    _mapView = mapView;
    
    NSArray<VisibleLayer*>* visibleLayers = [self defaultVisibleMapLayers];
    
    NSMutableArray<NSString*>* layerIdsUpdated = [NSMutableArray array];
    for (Filter* filter in _filters) {
        for (VisibleLayer* visibleLayer in visibleLayers) {
            if ([filter doFilterWithLayerId:visibleLayer.layerId type:visibleLayer.type sourceLayer:visibleLayer.sourceLayer]) {
                [layerIdsUpdated addObject:visibleLayer.layerId];
            }
        }
    }
    // Reset layers those are not effected.
    NSArray<MLNStyleLayer*>* layers = [_mapView.style layers];
    NSAssert([layers count], @"Empty layers???");
    for (VisibleLayer* visibleLayer in visibleLayers) {
        if (![layerIdsUpdated containsObject:visibleLayer.layerId]) {
            MLNStyleLayer* layer =  [layers findObjectUsingBlock:^BOOL(MLNStyleLayer* it) {
                return [it.identifier isEqualToString:visibleLayer.layerId];
            }];
            NSAssert(layer, @"Could not found the layer?");
            if ([layer isKindOfClass:MLNVectorStyleLayer.class]) {
                MLNVectorStyleLayer* vectorLayer = (MLNVectorStyleLayer*)layer;
                id expression = visibleLayer.filterExpression ? visibleLayer.filterExpression : [NSNumber numberWithBool:true];
                [vectorLayer setFilterExpression:expression];
            } else {
                [layer setVisible:YES];
            }
        }
    }

    _mapView = nil;
}

static NSMutableArray<VisibleLayer*>* _defaultVisibleMapLayers = nil;

-(NSArray<VisibleLayer*>*)defaultVisibleMapLayers {
    if (!_mapView) {
        @throw [NSException exceptionWithName:@"Failed to get visible map layers" reason:@"The map instance not been set" userInfo:nil];
    }
    
    if (!_defaultVisibleMapLayers) {
        NSString* backgroundLayerId = @"microsoft.bing.maps.base.land";
        
        _defaultVisibleMapLayers = [NSMutableArray array];
        NSArray<MLNStyleLayer*>* layers = [_mapView.style layers];
        if ([layers count]) {
            for (MLNStyleLayer* layer in layers) {
                NSString* layerId = layer.identifier;
                if (![layerId isEqualToString:backgroundLayerId] && layer.visible) {
                    NSString* type = @"rastor";
                    NSString* sourceLayer = @"";
                    id filterExpression = nil;
                    if ([layer isKindOfClass:MLNVectorStyleLayer.class]) {
                        MLNVectorStyleLayer* vectorLayer = (MLNVectorStyleLayer*)layer;
                        type = @"vector";
                        sourceLayer = vectorLayer.sourceLayerIdentifier;
                        filterExpression = vectorLayer.filterExpression;
                    }
                    [_defaultVisibleMapLayers addObject:[VisibleLayer layerWithLayerId:layerId type:type sourceLayer:sourceLayer filterExpression:filterExpression]];
                    NSLog(@"VisibleLayer: id:%@, type:%@, sourceLayer:%@, filterExpression:%@", layerId, type, sourceLayer, filterExpression);
                    
                }
            }
        }
    }
    return _defaultVisibleMapLayers;
}

+(void)resetAfterStyleChanged {
    _defaultVisibleMapLayers = nil;
}

@end

 
@implementation FilterCondition (NamedConditions)

static NSDictionary<NSString*, NSArray<FilterCondition*>*>* _namedConditions = nil;

+(NSDictionary<NSString*, NSArray<FilterCondition*>*>*) namedConditions {
    static dispatch_once_t tk;
    dispatch_once(&tk, ^{
        _namedConditions = @{
            @"default": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing.maps\\..*"]
            ],
            @"land": @[
                [FilterCondition conditionWithPattern:@"microsoft.bing.maps.baseFeature.vector_land"]
            ],
            @"base": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing.maps\\.baseFeature\\.([\\w\\d\\-_]+_fill|vector_land);"]
            ],
            @"reserve": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.baseFeature\\.[\\w\\d\\-_]+;reserve|golf_course"]
            ],
            @"hillShading": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.hillShading\\.hillShading;"]
            ],
            @"water": @[
                [FilterCondition conditionWithPattern:@"microsoft.bing.maps.baseFeature.generic_water_feature_fill"]
            ],
            @"sea": @[
                [FilterCondition conditionWithPattern:@"microsoft.bing.maps.baseFeature.generic_water_feature_fill" expression:@"[\"!\",[\"has\",\"st-et\"]]"]
            ],
            @"lake": @[
                [FilterCondition conditionWithPattern:@"microsoft.bing.maps.baseFeature.generic_water_feature_fill" expression:@"[\"==\",[\"get\",\"st-et\"],\"lake\"]"]
            ],
            @"river": @[
                [FilterCondition conditionWithPattern:@"microsoft.bing.maps.baseFeature.generic_water_feature_fill" expression:@"[\"==\",[\"get\",\"st-et\"],\"river\"]"]
            ],
            @"waterName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_water_feature_(labelonly|polygonlabel);"]
            ],
            @"seaName": @[
                [FilterCondition conditionWithPattern:@"microsoft.bing.maps.labels.generic_water_feature_labelonly;"]
            ],
            @"lakeName": @[
                [FilterCondition conditionWithPattern:@"microsoft.bing.maps.labels.generic_water_feature_polygonlabel" expression:@"[\"==\",[\"get\",\"st-et\"],\"lake\"]"]
            ],
            @"riverName": @[
                [FilterCondition conditionWithPattern:@"microsoft.bing.maps.labels.generic_water_feature_polygonlabel" expression:@"[\"==\",[\"get\",\"st-et\"],\"river\"]"]
            ],
            @"continentName": @[
                [FilterCondition conditionWithPattern:@"microsoft.bing.maps.labels.entity_override_continents_for_cn_region_symbol_label"]
            ],
            @"countryRegion": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.roads\\.[\\w\\d\\-_]+;country_region"]
            ],
            @"countryRegionName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.[\\w\\d\\-_]+;country_region"]
            ],
            @"islandName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.[\\w\\d\\-_]+;island"]
            ],
            @"cityName": @[
                // 首都
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_(sov_capital|beijing)_(labelonly|iconlabel);"],
                // 一般地名
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)"],
                // 乡镇
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_neighborhood_labelonly;"],
                // 台北
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_taipei_(labelonly|iconlabel)"],
                // 桃园
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_taoyuan_(labelonly|iconlabel)"],
                // 新北
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_new_taipei_(labelonly|iconlabel)"],
                // 其它增补城市（已确定：丹东、东港市、二连浩特）
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_label_orientation_[\\w\\d\\-_]+_(labelonly|iconlabel)"],
            ],
            @"capitalCityName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_(sov_capital|beijing)_(labelonly|iconlabel);"]
            ],
            @"admin1CityName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)" expression:@"[\"==\",[\"get\",\"cn-ppl\"],\"admin1cap\"]"],
                // 台北市
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_taipei_(labelonly|iconlabel)"],
            ],
            @"admin2CityName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)" expression:@"[\"==\",[\"get\",\"cn-ppl\"],\"admin2cap\"]"],
                // 桃园
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_taoyuan_(labelonly|iconlabel)"],
                // 新北
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_new_taipei_(labelonly|iconlabel)"],
                // 其它增补城市（已确定：丹东）
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_label_orientation_[\\w\\d\\-_]+_(labelonly|iconlabel)" expression:@"[\"==\",[\"get\",\"cn-ppl\"],\"admin2cap\"]"],
            ],
            // 一般城市（境外的直辖市、大城市等. NOTE: 国内无这个分类!）
            @"majorCityName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)" expression:@"[\"==\",[\"get\",\"cn-ppl\"],\"city\"]"],
                // 其它增补城市（已确定：东港市、二连浩特）
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_label_orientation_[\\w\\d\\-_]+_(labelonly|iconlabel)" expression:@"[\"==\",[\"get\",\"cn-ppl\"],\"city\"]"],
            ],
            @"minorCityName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)" expression:@"[\"==\",[\"get\",\"cn-ppl\"],\"minorcity\"]"],
                // 其它增补城市（已确定：东港市、二连浩特）
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_label_orientation_[\\w\\d\\-_]+_(labelonly|iconlabel)" expression:@"[\"==\",[\"get\",\"cn-ppl\"],\"minorcity\"]"],
            ],
            @"townName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_neighborhood_labelonly;"],
            ],
            @"villageName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.generic_populated_place_(labelonly|iconlabel)" expression:@"[\"==\",[\"get\",\"cn-ppl\"],\"village\"]"],
            ],
            @"road": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.roads\\.[\\w\\d\\-_]+;road[\\w\\d\\-_]*|railway[\\w\\d\\-_]*|tramway"],
            ],
            @"railway": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.roads\\.[\\w\\d\\-_]+;road[\\w\\d\\-_]*|railway[\\w\\d\\-_]*|tramway"],
            ],
            @"roadName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.[\\w\\d\\-_]+;road[\\w\\d\\-_]*|railway[\\w\\d\\-_]*|tramway"],
            ],
            @"railwayName": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\.[\\w\\d\\-_]+;railway[\\w\\d\\-_]*"],
            ],
            @"roadShield": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\..+_road_shield.*;road"],
            ],
            @"landmark": @[
                [FilterCondition conditionWithPattern:@"microsoft\\.bing\\.maps\\.labels\\..+_landmark(_\\w+)?"],
            ],
        };
    });
    return _namedConditions;
}
@end

